File: 213.House-Robber-II/dp.py
class Solution:
    def rob(self, nums: list[int]) -> int:
        n = len(nums)

        if n == 1:
            return nums[0]

        def dp(a, b):
            prev, curr = 0, nums[a]

            # O(n)
            for i in range(a + 2, b + 1):
                curr, prev = max(curr, prev + nums[i - 1]), curr

            return curr

        # O(n)
        return max(dp(1, n), dp(0, n - 1))

File: 45.Jump-Game-II/42.Trapping-Rain-Water/dp.rs
impl Solution {
    pub fn trap(height: Vec<i32>) -> i32 {
        // O(n)
        let left = height.iter().scan(0, |state, &x| {
            let prev = *state;
            *state = prev.max(x);
            Some(prev)
        });

        // O(n)
        let right = height
            .iter()
            .rev()
            .scan(0, |state, &x| {
                let next = *state;
                *state = next.max(x);
                Some(next)
            })
            .collect::<Vec<_>>()
            .into_iter()
            .rev();

        // O(n)
        height
            .iter()
            .zip(left.zip(right))
            .fold(0, |acc, (h, (l, r))| acc + (l.min(r) - h).max(0))
    }
}
